pkg/urbit/vere/io/khan.c

```diff
@@ -0,0 +1,477 @@
+      0,
+      0,
+      _khan_poke_bail);
+  u3_newt_moat_stop((u3_moat*)&can_u->mor_u, _khan_moat_free);
+}
+
+/* _khan_moor_bail(): error callback for u3_moor.
+*/
+static void
+_khan_moor_bail(void* ptr_v, ssize_t err_i, const c3_c* err_c)
+{
+  u3_chan*  can_u = (u3_chan*)ptr_v;
+  u3_shan*  san_u = can_u->san_u;
+  u3_chan*  inn_u;
+
+  if ( err_i == UV_EOF ) {
+    // close socket and remove reference.
+    if ( san_u->can_u == can_u ) {
+      san_u->can_u = (u3_chan*)can_u->mor_u.nex_u;
+    }
+    else {
+      for ( inn_u = san_u->can_u; inn_u; inn_u = (u3_chan*)inn_u->mor_u.nex_u ) {
+        if ( (u3_chan*)inn_u->mor_u.nex_u == can_u ) {
+          inn_u->mor_u.nex_u = can_u->mor_u.nex_u;
+          break;
+        }
+      }
+    }
+    can_u->mor_u.nex_u = NULL;
+    _khan_close_socket(can_u->san_u->kan_u, can_u);
+  }
+  else {
+    // TODO retry up to N
+    u3_noun bal;
+    c3_y*   byt_y;
+    c3_d    len_d;
+
+    bal = u3nq(c3__bail,
+               u3i_string("driver"),
+               -err_i & 0x7fffffff,
+               u3i_string(err_c));
+    u3s_jam_xeno(bal, &len_d, &byt_y);
+    u3_newt_send((u3_mojo*)&can_u->mor_u, len_d, byt_y);
+    u3z(bal);
+  }
+}
+
+/* _khan_moor_poke(): called on message read from u3_moor.
+*/
+static void
+_khan_moor_poke(void* ptr_v, c3_d len_d, c3_y* byt_y)
+{
+  u3_weak   jar;
+  u3_chan*  can_u = (u3_chan*)ptr_v;
+  u3_khan*  kan_u = can_u->san_u->kan_u;
+  u3_noun   wir;
+  u3_noun   cad;
+
+  jar = u3s_cue_xeno_with(kan_u->sil_u, len_d, byt_y);
+  if ( u3_none == jar ) {
+    can_u->mor_u.bal_f(can_u, -1, "cue-none");
+  }
+  else {
+    wir = u3nq(c3__khan,
+               u3dc("scot", c3__uv, kan_u->sev_l),
+               u3dc("scot", c3__ud, can_u->coq_l),
+               u3_nul);
+    cad = u3nc(c3__fyrd, jar);
+    u3_auto_peer(
+        u3_auto_plan(&kan_u->car_u, u3_ovum_init(0, c3__k, wir, cad)),
+        0,
+        0,
+        _khan_poke_bail);
+  }
+}
+
+/* _khan_conn_cb(): socket connection callback.
+*/
+static void
+_khan_conn_cb(uv_stream_t* sem_u, c3_i tas_i)
+{
+  u3_shan*  san_u = (u3_shan*)sem_u;
+  u3_khan*  kan_u = san_u->kan_u;
+  u3_chan*  can_u;
+  c3_i      err_i;
+
+  can_u = c3_calloc(sizeof(u3_chan));
+  can_u->mor_u.ptr_v = can_u;
+  can_u->mor_u.pok_f = _khan_moor_poke;
+  can_u->mor_u.bal_f = _khan_moor_bail;
+  can_u->coq_l = san_u->nex_l++;
+  can_u->san_u = san_u;
+  err_i = uv_timer_init(u3L, &can_u->mor_u.tim_u);
+  c3_assert(!err_i);
+  err_i = uv_pipe_init(u3L, &can_u->mor_u.pyp_u, 0);
+  c3_assert(!err_i);
+  err_i = uv_accept(sem_u, (uv_stream_t*)&can_u->mor_u.pyp_u);
+  c3_assert(!err_i);
+  u3_newt_read((u3_moat*)&can_u->mor_u);
+  can_u->mor_u.nex_u = (u3_moor*)san_u->can_u;
+  san_u->can_u = can_u;
+}
+
+/* _khan_sock_init(): initialize socket device.
+*/
+static void
+_khan_sock_init(u3_shan* san_u)
+{
+  // The full socket path is limited to about 108 characters, and we want it to
```

Comment:
I see this protocol is intended for control backplane in hosted scenarios, so idk if it should exist on Windows at all, but just in case - the closest equivalent of Unix sockets on Windows, named pipes, do not share a namespace with the filesystem and cannot be linked/mounted into it relative to anything. Easiest way is probably to create a pipe name based on ship name, something like `\.\pipe\%khan-sampel-palnet`.

Reply:
it's first use case is for that, but it will definitely not be only for that. it should certainly exist on windows.

Reply:
for now i'm no-opping on windows; happy to come back to this and make pipes work later.